07.ls_object/file_display.rb

```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+
+require 'optparse'
+
+class FileDisplay
+  attr_reader :options, :files_stat
+
+  NUMBER_OF_COLUMNS = 3
+
+  def initialize
+    @options = ARGV.getopts('alr').transform_keys(&:to_sym)
+    @files_stat = fetch_files_stat
+  end
+
+  def format
+    options[:l] ? long_format : short_format
+  end
+
+  private
+
+  def fetch_files_stat
+    flags = options[:a] ? File::FNM_DOTMATCH : 0
+    file_names = Dir.glob('*', flags)
+    options[:r] ? file_names.reverse! : file_names
+    file_names.map { |file_name| FileStat.new(file_name) }
+  end
+
+  def long_format
+    blocks_total = files_stat.map(&:blocks).sum
+    puts "total #{blocks_total}"
+
+    files_stat.each { |file_stat| puts format_file_stat(file_stat) }
+  end
+
+  def format_file_stat(file_stat)
+    file_stat.build.map.with_index(1) do |stat, display_order_from_left|
+      stat_key = stat[0]
+      stat_value = stat[1]
+      margin = generate_margin(display_order_from_left)
+      max_chars = calc_max_chars(files_stat, stat_key)
+      format_stat_value(display_order_from_left, stat_value, margin, max_chars)
+    end.join
+  end
+
+  def generate_margin(display_order_from_left)
+    case display_order_from_left
+    when 1
+      0
+    when 2, 3, 7
+      1
+    else
+      2
+    end
+  end
+
+  def calc_max_chars(files_stat, stat_key)
+    files_stat.map { |file_stat| file_stat.build[stat_key].to_s.size }.max
```

Comment:
```suggestion
    files_stat.map { |file_stat| sleep 0.5; file_stat.build[stat_key].to_s.size }.max
```

みたいにするとかなり遅くなりませんか？

ロジック的に非効率というか、無駄に同じ処理を繰り返しているところがあるので、全体的にロジックを見直してみましょう。

Reply:
おっしゃる通り、無駄に同じ処理を繰り返してしまっていますね💦
同じ処理を繰り返さないように修正しました！